src/Stratis.Features.FederatedPeg/TargetChain/SigningUtils.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using NBitcoin;
+using Stratis.Bitcoin.Utilities;
+using Stratis.Features.FederatedPeg.Interfaces;
+
+namespace Stratis.Features.FederatedPeg.TargetChain
+{
+    /// <summary>
+    /// Exposes methods that help with the identifying and signing of transactions deterministically.
+    /// </summary>
+    public static class SigningUtils
+    {
+        public static int GetSignatureCount(this Transaction partialTransaction, Network network)
+        {
+            Guard.NotNull(partialTransaction, nameof(partialTransaction));
+            Guard.Assert(partialTransaction.Inputs.Any());
+
+            Script scriptSig = partialTransaction.Inputs[0].ScriptSig;
+            if (scriptSig == null)
+                return 0;
+
+            // Remove the script from the end.
+            scriptSig = new Script(scriptSig.ToOps().SkipLast(1));
+
+            TransactionSignature[] result = PayToMultiSigTemplate.Instance.ExtractScriptSigParameters(network, scriptSig);
+
+            return result?.Count(s => s != null) ?? 0;
+        }
+
+        public static Transaction CombineSignatures(TransactionBuilder builder, Transaction existingTransaction, Transaction[] partialTransactions)
+        {
+            Transaction[] validPartials = partialTransactions.Where(p => TemplatesMatch(builder.Network, p, existingTransaction) && p.GetHash() != existingTransaction.GetHash()).ToArray();
+            if (validPartials.Any())
+            {
+                var allPartials = new Transaction[validPartials.Length + 1];
+                allPartials[0] = existingTransaction;
+                validPartials.CopyTo(allPartials, 1);
+
+                existingTransaction = builder.CombineSignatures(allPartials);
+            }
+
+            return existingTransaction;
+        }
+
+
+        /// <summary>
+        /// Checks whether two transactions have identical inputs and outputs.
+        /// </summary>
+        /// <param name="partialTransaction1">First transaction.</param>
+        /// <param name="partialTransaction2">Second transaction.</param>
+        /// <returns><c>True</c> if identical and <c>false</c> otherwise.</returns>
+        public static bool TemplatesMatch(Network network, Transaction partialTransaction1, Transaction partialTransaction2)
```

Comment:
What's the significance of "template"? Why not something like "IsTheSameAs" or "IsEqualTo'?

Reply:
Good question - I moved this out from CrossChainTransfer so it could be used in more places, I didn't decide the naming myself but you raise a good point. Future PR?

Reply:
I think we called the partials 'template transactions' in early versions of the cross chain transfer design; it could certainly be renamed.

Reply:
Future PR then :)